libpod/container_internal_linux.go

```diff
@@ -696,3 +697,79 @@ func (c *Container) getConmonPidFd() int {
 	}
 	return -1
 }
+
+type safeMountInfo struct {
+	// file is the open File.
+	file *os.File
+
+	// mountPoint is the mount point.
+	mountPoint string
+}
+
+// Close releases the resources allocated with the safe mount info.
+func (s *safeMountInfo) Close() {
+	_ = unix.Unmount(s.mountPoint, unix.MNT_DETACH)
+	_ = s.file.Close()
+}
+
+// safeMountSubPath securely mounts a subpath inside a volume to a new temporary location.
+// The function checks that the subpath is a valid subpath within the volume and that it
+// does not escape the boundaries of the mount point (volume).
+//
+// The caller is responsible for closing the file descriptor and unmounting the subpath
+// when it's no longer needed.
+func (c *Container) safeMountSubPath(mountPoint, subpath string) (s *safeMountInfo, err error) {
+	joinedPath := filepath.Clean(filepath.Join(mountPoint, subpath))
+	fd, err := unix.Open(joinedPath, unix.O_RDONLY|unix.O_PATH, 0)
+	if err != nil {
+		return nil, err
+	}
+	f := os.NewFile(uintptr(fd), joinedPath)
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	fdPath := fmt.Sprintf("/proc/%d/fd/%d", os.Getpid(), f.Fd())
```

Comment:
Do I understand it correctly that the trick to avoid a TOCTOU is by using the FD here?  The path may change but the FD will continue pointing to the same file?

Whether I understood it correctly or not, I think that a comment would be nice.

Reply:
the trick is that once it is opened, it cannot become a symlink anymore.  So if we open, validate it and then use it, we are safe.  I'll add a comment

Reply:
Excellent, thanks!

Clever idea, btw